hack/generator/pkg/jsonast/schema_abstraction_openapi.go

```diff
@@ -0,0 +1,288 @@
+}
+
+func (schema *OpenAPISchema) requiredProperties() []string {
+	return schema.inner.Required
+}
+
+func (schema *OpenAPISchema) properties() map[string]Schema {
+	result := make(map[string]Schema)
+	for propName, propSchema := range schema.inner.Properties {
+		result[propName] = schema.withNewSchema(propSchema)
+	}
+
+	return result
+}
+
+func (schema *OpenAPISchema) description() *string {
+	if len(schema.inner.Description) == 0 {
+		return nil
+	}
+
+	return &schema.inner.Description
+}
+
+func (schema *OpenAPISchema) items() []Schema {
+	if schema.inner.Items.Schema != nil {
+		return []Schema{schema.withNewSchema(*schema.inner.Items.Schema)}
+	}
+
+	return schema.transformOpenAPISlice(schema.inner.Items.Schemas)
+}
+
+func (schema *OpenAPISchema) additionalPropertiesAllowed() bool {
+	return schema.inner.AdditionalProperties == nil || schema.inner.AdditionalProperties.Allows
+}
+
+func (schema *OpenAPISchema) additionalPropertiesSchema() Schema {
+	if schema.inner.AdditionalProperties == nil {
+		return nil
+	}
+
+	result := schema.inner.AdditionalProperties.Schema
+	if result == nil {
+		return nil
+	}
+
+	return schema.withNewSchema(*result)
+}
+
+func (schema *OpenAPISchema) enumValues() []string {
+	result := make([]string, len(schema.inner.Enum))
+	for i, enumValue := range schema.inner.Enum {
+		if enumString, ok := enumValue.(string); ok {
+			result[i] = fmt.Sprintf("%q", enumString)
+		} else if enumStringer, ok := enumValue.(fmt.Stringer); ok {
+			result[i] = fmt.Sprintf("%q", enumStringer.String())
+		} else if enumFloat, ok := enumValue.(float64); ok {
+			result[i] = fmt.Sprintf("%g", enumFloat)
+		} else {
+			panic(fmt.Sprintf("unable to convert enum value (%v %T) to string", enumValue, enumValue))
+		}
+	}
+
+	return result
+}
+
+func (schema *OpenAPISchema) isRef() bool {
+	return schema.inner.Ref.GetURL() != nil
+}
+
+// fetchFileRelative fetches the schema for the relative path created by combining 'baseFileName' and 'url'
+func (fileCache *OpenAPISchemaCache) fetchFileRelative(baseFileName string, url *url.URL) (string, spec.Swagger, error) {
+	path := ""
+	swagger := spec.Swagger{}
+
+	if url.IsAbs() {
+		return path, swagger, errors.Errorf("only relative URLs can be handled")
+	}
+
+	fileURL, err := url.Parse("file://" + baseFileName)
+	if err != nil {
+		return path, swagger, errors.Wrapf(err, "cannot convert filename to file URI")
+	}
+
+	path = fileURL.ResolveReference(url).Path
+	swagger, err = fileCache.fetchFileAbsolute(path)
+
+	return path, swagger, err
+}
+
+// fetchFileAbsolute fetches the schema for the absolute path specified
+func (fileCache *OpenAPISchemaCache) fetchFileAbsolute(filePath string) (spec.Swagger, error) {
+	if swagger, ok := fileCache.files[filePath]; ok {
+		return swagger, nil
+	}
+
+	var swagger spec.Swagger
+
+	fileContent, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return swagger, errors.Wrap(err, "unable to read swagger file")
```

Comment:
Same question as before about needing file paths here in the errors?

Also, worth factoring out to a little helper function since there seem to be 2 places that do the same thing w/ the same errors?

Reply:
I had thought about a common function but they live in different modules and the "sharing" would be one module reaching into another's internal helper functions. I don't think the duplication is a big deal here.

Reply:
Ah, I missed that they were in different modules. Carry on then, fine as is.